### PR TITLE
Flags: remove protocol sniffing flags

### PR DESCRIPTION
**Pull request by howardjohn**
This has been enabled by default since 1.5 and is now a part of core Istio. Time to clean up all the dead code we have to support this flag. There will be more followup cleanup that this _enables_ but is not really realted to protocol sniffing so I left it ouut

TODO:
* Look for other places
* Add release notes
* Fix tests, probably

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
I am sure we have some clusters still disable sniff

**Comment by howardjohn** [comment]
Can you help understand why?

This should either be a proper Istio feature and API or removed eventually
imo

**Comment by hzxuzhonghu** [comment]
IIRC, latency is the most critical point.

**Comment by howardjohn** [comment]
Is there any real data on this? When Istio released the feature it was
measured and there was no performance overhead at the time?

**Comment by hzxuzhonghu** [comment]
Hmm, let me ask relevant people about this. Protocol sniff should have some cost on inspector

**Comment by howardjohn** [comment]
It could theoretically but in practice it's just reading first few bytes of
connection, very very little overhead in the testing I have seen. Also we
usually are also sniffing TLS which is more complex

**Comment by howardjohn** [comment]
@hzxuzhonghu fyi, this will enable another PR  I am working on:

```
name                       old time/op        new time/op        delta
ListenerGeneration/http-8         123µs ±23%          99µs ±13%  -19.74%  (p=0.000 n=10+24)
ListenerGeneration/tcp-8          537µs ±24%         372µs ± 4%  -30.79%  (p=0.000 n=10+23)
ListenerGeneration/tls-8          550µs ±12%         386µs ± 5%  -29.77%  (p=0.000 n=10+25)
ListenerGeneration/auto-8        1.96ms ±13%        1.73ms ± 5%  -11.88%  (p=0.002 n=10+20)

name                       old alloc/op       new alloc/op       delta
ListenerGeneration/http-8         109kB ± 0%          74kB ± 0%  -32.20%  (p=0.000 n=10+25)
ListenerGeneration/tcp-8          424kB ± 0%         363kB ± 0%  -14.43%  (p=0.000 n=10+24)
ListenerGeneration/tls-8          425kB ± 0%         364kB ± 0%  -14.42%  (p=0.000 n=10+24)
ListenerGeneration/auto-8        1.44MB ± 0%        1.34MB ± 0%   -7.15%  (p=0.000 n=8+24)

name                       old allocs/op      new allocs/op      delta
ListenerGeneration/http-8         1.20k ± 0%         0.88k ± 0%  -26.11%  (p=0.000 n=10+25)
ListenerGeneration/tcp-8          5.98k ± 0%         4.38k ± 0%  -26.82%  (p=0.000 n=9+24)
ListenerGeneration/tls-8          5.99k ± 0%         4.38k ± 0%  -26.80%  (p=0.000 n=10+25)
ListenerGeneration/auto-8         16.7k ± 0%         14.7k ± 0%  -12.02%  (p=0.000 n=8+25)
```

Benchmark is not from this change (this one should have no impact), but it is enabled by the simplification here.

**Comment by hzxuzhonghu** [comment]
@howardjohn Asked about this, we have disabled protocol sniffer on a very old version maybe 1.8. In recent versions, we enabled this always, so lgtm

**Comment by howardjohn** [comment]
great, thank you for checking!

File: pilot/pkg/networking/core/v1alpha3/cluster_test.go [hunk 1]
```diff
@@ -147,46 +147,40 @@ func TestHTTPCircuitBreakerThresholds(t *testing.T) {
...
-		sniffingEnabledForInbound bool
-		proxyType                 model.NodeType
-		clusters                  int
+		clusterName           string
+		useDownStreamProtocol bool
+		proxyType             model.NodeType
+		clusters              int
 	}{
 		{
-			clusterName:               "outbound|8080||*.example.org",
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster_test.go:157]: nit: sniffingEnabledForInbound -> false tests can be removed?

**Review by ramaraochavali — COMMENTED** [review]
Nice cleanup. One test comment - LGTM

File: pilot/pkg/networking/core/v1alpha3/cluster_test.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_test.go:157]: I don't think we have any duplicate coverage now. Just had less coverage before

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster_test.go:157]: All cases are unique combination of direction, useDownStreamProtocol, and proxyType

**Review by ramaraochavali — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 2]
```diff
@@ -713,9 +713,9 @@ message PilotConfig {
   repeated google.protobuf.Struct tolerations = 26 [deprecated=true];
 
   // if protocol sniffing is enabled for outbound
-  google.protobuf.BoolValue enableProtocolSniffingForOutbound = 28;
+  google.protobuf.BoolValue enableProtocolSniffingForOutbound = 28 [deprecated=true];
   // if protocol sniffing is enabled for inbound
-  google.protobuf.BoolValue enableProtocolSniffingForInbound = 29;
+  google.protobuf.BoolValue enableProtocolSniffingForInbound = 29 [deprecated=true];
```
Comment (hzxuzhonghu) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto:718]: It is removed, so can we remove the field here

**Pull request merged**